unidist/core/backends/mpi/core/communication.py

```diff
@@ -191,27 +246,78 @@ def send_complex_data(comm, data, dest_rank):
         Data object to send.
     dest_rank : int
         Target MPI process to transfer data.
+
+    Returns
+    -------
+    object
+        A serialized msgpack data.
+    list
+        A list of pickle buffers.
+    list
+        A list of buffers amount for each object.
     """
-    serializer = MPISerializer()
+    serializer = ComplexSerializer()
     # Main job
     s_data = serializer.serialize(data)
-
-    # Send message pack bytestring
-    comm.send(len(s_data), dest=dest_rank)
-    comm.Send([s_data, MPI.CHAR], dest=dest_rank)
     # Retrive the metadata
     raw_buffers = serializer.buffers
     len_buffers = serializer.len_buffers
 
+    # MPI comminucation
+    _send_complex_data_impl(comm, s_data, raw_buffers, len_buffers, dest_rank)
+
+    # For caching purpose
+    return s_data, raw_buffers, len_buffers
+
+
+def _isend_complex_data_impl(comm, s_data, raw_buffers, len_buffers, dest_rank):
+    """
+    Send serialized complex data.
+
+    Non-blocking asynchronous interface.
+
+    Parameters
+    ----------
+    comm : object
+        MPI communicator object.
+    s_data: object
+        A serialized msgpack data.
+    raw_buffers : list
+        A list of pickle buffers.
+    len_buffers : list
+        A list of buffers amount for each object.
+    dest_rank : int
+        Target MPI process to transfer data.
+
+    Returns
+    -------
+    list
+        A list of pairs, ``MPI_Isend`` handler and associated data to send.
+    """
+    handlers = []
+
+    # Send message pack bytestring
+    h1 = mpi_isend_object(comm, len(s_data), dest_rank)
+    h2 = mpi_isend_buffer(comm, s_data, dest_rank)
+    handlers.append((h1, None))
+    handlers.append((h2, s_data))
+
     # Send the necessary metadata
-    comm.send(len(raw_buffers), dest=dest_rank)
-    for i in range(0, len(raw_buffers)):
-        comm.send(len(raw_buffers[i].raw()), dest=dest_rank)
-        comm.Send([raw_buffers[i], MPI.CHAR], dest=dest_rank)
-    comm.send(len_buffers, dest=dest_rank)
+    h3 = mpi_isend_object(comm, len(raw_buffers), dest_rank)
+    handlers.append((h3, None))
+    for raw_buffer in raw_buffers:
+        h4 = mpi_isend_object(comm, len(raw_buffer.raw()), dest_rank)
+        h5 = mpi_isend_buffer(comm, raw_buffer, dest_rank)
+        handlers.append((h4, None))
+        handlers.append((h5, raw_buffer))
+    # TODO: do not send if raw_buffers is zero
+    h6 = mpi_isend_object(comm, len_buffers, dest_rank)
```

Comment:
There is at least 4 sends. Can we lessen these?

Reply:
Answered above